GoogleApiWrappers/src/Google.Apis.Storage.v1.ClientWrapper/StorageClient.cs

```diff
@@ -0,0 +1,105 @@
+        public StorageClient(StorageService service)
+        {
+            Service = service;
+        }
+
+        /// <summary>
+        /// Constructs a new client by creating a <see cref="StorageService"/> which uses the
+        /// given <see cref="BaseClientService.Initializer"/> for request initialization.
+        /// </summary>
+        /// <param name="initializer">The initializer to use in the service. Must not be null.</param>
+        public StorageClient(BaseClientService.Initializer initializer) : this(new StorageService(initializer))
+        {
+        }
+
+        /// <summary>
+        /// Asynchronously creates a new <see cref="StorageClient"/> from the default application credentials.
+        /// </summary>
+        /// <remarks>
+        /// The application credentials are available to developers by running <c>gcloud auth</c> from the
+        /// command line, and are available automatically on Google Cloud Platform hosts.
+        /// </remarks>
+        /// <param name="applicationName">The name of the application to create a client for. Must not be null.</param>
+        /// <returns>A client configured with the application-default credentials.</returns>
+        public static async Task<StorageClient> FromApplicationCredentials(string applicationName)
+        {
+            var initializer = new BaseClientService.Initializer
+            {
+                HttpClientInitializer = await GoogleCredential.GetApplicationDefaultAsync(),
+                ApplicationName = applicationName,
+            };
+            return new StorageClient(initializer);
+        }
+
+        /// <summary>
+        /// Asynchronously lists all the buckets for a given project.
+        /// </summary>
+        /// <remarks>
+        /// This is a convenience method for calling <see cref="ListAllBucketsAsync(string, CancellationToken)"/>.
+        /// </remarks>
+        /// <param name="project">The project to list the buckets from. Must not be null.</param>
+        public Task<List<Bucket>> ListAllBucketsAsync(string project)
+        {
+            return ListAllBucketsAsync(project, cancellationToken: default(CancellationToken));
+        }
+
+        /// <summary>
+        /// Asynchronously lists all the buckets for a given project.
+        /// </summary>
+        /// <remarks>
+        /// This lists all the buckets within a project before the returned task completes.
+        /// This does not support reporting progress, or streaming the results.
+        /// </remarks>
+        /// <param name="project">The project to list the buckets from. Must not be null.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        /// <returns>A list of all buckets within the project.</returns>
+        public async Task<List<Bucket>> ListAllBucketsAsync(string project, CancellationToken cancellationToken)
```

Comment:
Just thinking ahead, compatibility-wise we don't need to marry ourselves with the List<T> implementation only that it is list-like object vs. an enumerable-like object, the user doesn't need to know that it is a List<T> vs. our own class if we need that in the future.

About going down to IEnumerable<T> and reading lazily is probably the best way to go for the sync method actually, for the async, not sure how about how to impelemnt IAsyncEnumerator just yet... Perhaps looking into Rx for that?